### PR TITLE
Replace vpc controller with networkinfo controller

### PR DESCRIPTION
**Pull request by zhengxiexie**
Since force pushing to a closed PR https://github.com/vmware-tanzu/nsx-operator/pull/553, can't reopen it, thus creating a new PR, the comments stay the same with it.

### CODE REVIEW

File: pkg/controllers/namespace/namespace_controller.go [hunk 1]
```diff
@@ -247,7 +239,7 @@ func (r *NamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			}
 		}
 
-		if _, err := r.createVPCCR(&ctx, obj, ns, ncName, createVpcName); err != nil {
+		if _, err := r.createNetworkInfoCR(&ctx, obj, ns, ncName); err != nil {
 			return common.ResultRequeueAfter10sec, nil
 		}
 		if err := r.createDefaultSubnetSet(ns); err != nil {
```
Comment (seanpang-vmware) [inline pkg/controllers/namespace/namespace_controller.go:245]: just leave a question here, do we need to create default subnetset cr for shared vpc here

File: pkg/controllers/namespace/namespace_controller_test.go [hunk 2]
```diff
@@ -121,111 +109,3 @@ func TestInsertNamespaceNetworkconfigBinding(t *testing.T) {
 		})
 	}
 }
-
-func TestCreateVPCCR(t *testing.T) {
```
Comment (seanpang-vmware) [inline pkg/controllers/namespace/namespace_controller_test.go:125]: we need to add the unit test back when we have time...

File: pkg/nsx/services/ippool/ippool.go [hunk 3]
```diff
@@ -12,6 +12,7 @@ import (
 	"k8s.io/client-go/tools/cache"
 
 	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha2"
+
```
Comment (seanpang-vmware) [inline pkg/nsx/services/ippool/ippool.go:15]: unnecessary change

File: pkg/nsx/services/vpc/builder.go [hunk 4]
```diff
@@ -61,9 +65,12 @@ func buildNSXVPC(obj *v1alpha1.VPC, nc common.VPCNetworkConfigInfo, cluster stri
 		vpc = nsxVPC
 	} else {
 		// for creating vpc case, fill in vpc properties based on networkconfig
-		vpcName := util.GenerateDisplayName(obj.Name, obj.GetNamespace(), cluster, "", "")
+		vpcName := sharedVPCName
```
Comment (seanpang-vmware) [inline pkg/nsx/services/vpc/builder.go:68]: although functional work, but the logic order seems a little odd

File: pkg/nsx/services/vpc/vpc.go [hunk 5]
```diff
@@ -473,9 +477,44 @@ func (s *VPCService) CreateorUpdateVPC(obj *v1alpha1.VPC) (*model.Vpc, *common.V
...
+		} else {
+			// temp: find vmware-system-supervisor-services-vpc has anno nsx.vmware.com/vpc_name: vmware-system-supervisor-services-vpc/vmware-system-supervisor-services
+			// this mislead shared vpc is vmware-system-supervisor-services
+			if res[1] == "infrastructure" {
+				sharedVPCName = res[1]
+			}
+		}
+	}
+
+	existingVPC := s.VpcStore.GetVPCsByNamespace(ns)
```
Comment (seanpang-vmware) [inline pkg/nsx/services/vpc/vpc.go:509]: you could call this method directly, it already cover the scenario if this ns is using shared vpc.

File: .gitignore [hunk 6]
```diff
@@ -2,4 +2,5 @@
 bin/
 .DS_Store
 go.work
-go.work.sum
\ No newline at end of file
+.coverage
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply .gitignore:5]: ok

File: pkg/controllers/namespace/namespace_controller.go [hunk 7]
```diff
@@ -73,22 +71,27 @@ func (r *NamespaceReconciler) createVPCCR(ctx *context.Context, obj client.Objec
...
 	}
 
-	// create vpc cr with existing vpc network config
-	vpcCR := BuildVPCCR(ns, ncName, vpcName)
-	err := r.Client.Create(*ctx, vpcCR)
+	// create networkInfo cr with existing vpc network config
+	log.V(2).Info("building networkInfo", "ns", ns)
+	networkInfoCR := &v1alpha1.NetworkInfo{}
+	networkInfoCR.Name = ns
+	networkInfoCR.Namespace = ns
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/namespace/namespace_controller.go:78]: ok

File: pkg/controllers/namespace/namespace_controller.go (hunk 1 above)
Comment (zhengxiexie) [reply pkg/controllers/namespace/namespace_controller.go:245]: as discussed

File: pkg/controllers/namespace/namespace_controller_test.go (hunk 2 above)
Comment (zhengxiexie) [reply pkg/controllers/namespace/namespace_controller_test.go:125]: ok

File: pkg/nsx/services/ippool/ippool.go (hunk 3 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ippool/ippool.go:15]: ok

File: pkg/nsx/services/vpc/vpc.go (hunk 5 above)
Comment (zhengxiexie) [reply pkg/nsx/services/vpc/vpc.go]: fixed

File: pkg/nsx/services/vpc/builder.go (hunk 4 above)
Comment (zhengxiexie) [reply pkg/nsx/services/vpc/builder.go:68]: could removed as discussed

**Review by seanpang-vmware — DISMISSED** [review]

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 8]
```diff
@@ -27,52 +27,50 @@ import (
...
+				log.Error(err, "add finalizer", "NetworkInfo", req.NamespacedName)
 				updateFail(r, &ctx, obj, &err, r.Client)
 				return common.ResultRequeue, err
 			}
-			log.V(1).Info("added finalizer on VPC CR", "VPC", req.NamespacedName)
+			log.V(1).Info("added finalizer on NetworkInfo CR", "NetworkInfo", req.NamespacedName)
 		}
 
-		createdVpc, nc, err := r.Service.CreateorUpdateVPC(obj)
+		createdVpc, nc, err := r.Service.CreateOrUpdateVPC(obj)
```
Comment (TaoZou1) [inline pkg/controllers/networkinfo/networkinfo_controller.go:65]: CreateOrUpdateAVIRule should be skipped if createdVpc returned from CreateOrUpdateVPC is a shared VPC(not created for this ns), check IsSharedVPCNamespaceByNS before call CreateOrUpdateAVIRule.

File: pkg/nsx/services/vpc/vpc.go [hunk 9]
```diff
@@ -330,6 +330,20 @@ func (s *VPCService) CreatOrUpdatePrivateIPBlock(obj *v1alpha1.VPC, nc common.VP
 	return path, nil
 }
 
+func (s *VPCService) IsSharedVPCNamespaceByNS(ns string) (bool, error) {
+	shared_ns, err := s.getSharedVPCNamespaceFromNS(ns)
```
Comment (TaoZou1) [inline pkg/nsx/services/vpc/vpc.go:334]: could we use the vpc.tags to check if it's a shared VPC instead of using a getSharedVPCNamespaceFromNS ?

File: pkg/controllers/networkinfo/networkinfo_controller.go (hunk 8 above)
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go:65]: fixed

File: pkg/nsx/services/vpc/vpc.go (hunk 9 above)
Comment (zhengxiexie) [reply pkg/nsx/services/vpc/vpc.go:334]: No, it is in annotation.

File: pkg/nsx/services/vpc/builder.go [hunk 10]
```diff
@@ -19,28 +20,28 @@ var (
...
 		}
 	}
-	return *cidr + "_" + vpc_uid
+	return *cidr + "_" + ns_uid
 }
 
-func generateIPBlockSearchKey(cidr string, vpcCRUID string) string {
-	return cidr + "_" + vpcCRUID
+func generateIPBlockSearchKey(cidr string, nsCRUID string) string {
+	return cidr + "_" + nsCRUID
```
Comment (heypnus) [inline pkg/nsx/services/vpc/builder.go:33]: `ns_uid`/`nsCRUID` -> `nsUID`

File: pkg/nsx/services/vpc/vpc.go [hunk 11]
```diff
@@ -260,13 +259,13 @@ func (s *VPCService) DeleteIPBlockInVPC(vpc model.Vpc) error {
...
+		nsCRUid := ""
 		for _, tag := range vpc.Tags {
-			if *tag.Scope == common.TagScopeVPCCRUID {
-				vpcCRUid = *tag.Tag
+			if *tag.Scope == common.TagScopeNamespaceUID {
+				nsCRUid = *tag.Tag
 			}
 		}
-		log.V(2).Info("search ip block from store using index and path", "index", common.TagScopeVPCCRUID, "Value", vpcCRUid, "Path", block)
+		log.V(2).Info("search ip block from store using index and path", "index", common.TagScopeNamespaceUID, "Value", nsCRUid, "Path", block)
```
Comment (heypnus) [inline pkg/nsx/services/vpc/vpc.go:268]: `nsCRUid` -> `nsUID`

File: pkg/nsx/services/vpc/vpc.go [hunk 12]
```diff
@@ -714,7 +752,7 @@ func (service *VPCService) buildAVIGroupTag(vpcId string) []model.Tag {
 			Tag:   common.String(strings.Join(common.TagValueVersion, ".")),
 		},
 		{
-			Scope: common.String(common.TagScopeVPCCRUID),
+			Scope: common.String(common.TagScopeNamespaceUID), // vpcId is ns uid
```
Comment (heypnus) [inline pkg/nsx/services/vpc/vpc.go:755]: This scope changed from `TagScopeVPCCRUID` to `TagScopeNamespaceUID`, is there any related change? i.e. where consumes this tag?

File: pkg/controllers/namespace/namespace_controller.go [hunk 13]
```diff
@@ -73,22 +71,30 @@ func (r *NamespaceReconciler) createVPCCR(ctx *context.Context, obj client.Objec
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ns,
+			Namespace: ns,
+		},
+		VPCs: []v1alpha1.VPCState{},
+	}
+	err := r.Client.Create(*ctx, networkInfoCR)
 	if err != nil {
-		message := "failed to create VPC CR"
+		message := "failed to create or update NetworkInfo CR"
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/namespace/namespace_controller.go:85]: ok

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 14]
```diff
@@ -99,16 +105,21 @@ func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
...
 		}
 
-		updateSuccess(r, &ctx, obj, r.Client, *createdVpc.Path, snatIP, path, cidr, nc.PrivateIPv4CIDRs)
+		updateSuccess(r, &ctx, obj, r.Client, *createdVpc.DisplayName, *createdVpc.Path, snatIP, path, cidr, nc.PrivateIPv4CIDRs, nc.Name)
 	} else {
-		if controllerutil.ContainsFinalizer(obj, commonservice.VPCFinalizerName) {
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, common.MetricResTypeVPC)
+		if controllerutil.ContainsFinalizer(obj, commonservice.NetworkInfoFinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, common.MetricResTypeNetworkInfo)
 			vpcs := r.Service.GetVPCsByNamespace(obj.GetNamespace())
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go:112]: ok

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 15]
```diff
@@ -99,16 +105,21 @@ func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
...
+			isShared, err := r.Service.IsSharedVPCNamespaceByNS(obj.GetNamespace())
+			if err != nil {
+				log.Error(err, "failed to check if namespace is shared", "Namespace", obj.GetNamespace())
+				return common.ResultRequeue, err
+			}
 			// if nsx resource do not exist, continue to remove finalizer, or the crd can not be removed
 			if len(vpcs) == 0 {
 				// when nsx vpc not found in vpc store, skip deleting NSX VPC
 				log.Info("can not find VPC in store, skip deleting NSX VPC, remove finalizer from VPC CR")
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go:121]: ok

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 16]
```diff
@@ -99,16 +105,21 @@ func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
...
-			} else {
+			} else if !isShared {
 				vpc := vpcs[0]
 				if err := r.Service.DeleteVPC(*vpc.Path); err != nil {
 					log.Error(err, "failed to delete VPC CR, would retry exponentially", "VPC", req.NamespacedName)
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go:125]: ok

File: pkg/controllers/networkinfo/networkinfo_utils.go [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+			client.Status().Update(*ctx, nc)
+			return
+		}
+	}
+	// else append the new VPCInfo
+	if nc.Status.VPCs == nil {
+		nc.Status.VPCs = []v1alpha1.VPCInfo{}
+	}
+	nc.Status.VPCs = append(nc.Status.VPCs, *createdVPCInfo)
+	client.Status().Update(*ctx, nc)
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_utils.go:88]: I see many unhandled error in `client.Update()` of all controllers, I think we can rearrange it in the future.

File: pkg/controllers/networkinfo/vpcnetworkconfig_handler.go [hunk 18]
```diff
@@ -66,22 +70,20 @@ func (h *VPCNetworkConfigurationHandler) Update(_ context.Context, e event.Updat
 	h.vpcService.RegisterVPCNetworkConfig(newNc.Name, *info)
 
 	nss := h.vpcService.GetNamespacesByNetworkconfigName(newNc.Name)
-	ctx := context.Background()
-	// find vpcs under each ns, and reconcile the vpc object
 	for _, ns := range nss {
-		vpcList := &v1alpha1.VPCList{}
-		err := h.Client.List(ctx, vpcList, client.InNamespace(ns))
+		networkInfos := &v1alpha1.NetworkInfoList{}
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/vpcnetworkconfig_handler.go:74]: No, only one.

File: pkg/nsx/services/vpc/builder.go (hunk 10 above)
Comment (zhengxiexie) [reply pkg/nsx/services/vpc/builder.go:33]: ok

File: pkg/nsx/services/vpc/vpc.go (hunk 11 above)
Comment (zhengxiexie) [reply pkg/nsx/services/vpc/vpc.go:268]: ok

File: pkg/nsx/services/vpc/vpc.go (hunk 12 above)
Comment (zhengxiexie) [reply pkg/nsx/services/vpc/vpc.go:755]: I think we can remove this tag, @TaoZou1 pls help confirm.

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 19]
```diff
@@ -136,9 +147,9 @@ func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 	return common.ResultNormal, nil
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go:147]: sorry?

File: pkg/nsx/services/vpc/vpc.go (hunk 12 above)
Comment (TaoZou1) [reply pkg/nsx/services/vpc/vpc.go]: yes

Comment (zhengxiexie) [reply pkg/nsx/services/vpc/vpc.go]: OK，removed

File: pkg/controllers/networkinfo/networkinfo_controller.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go:147]: ok

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 20]
```diff
@@ -121,7 +132,7 @@ func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 				}
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/controllers/networkinfo/networkinfo_controller.go]: fixed

**Review by seanpang-vmware — APPROVED** [review]
LGTM
